src/rmt.rs

```diff
@@ -0,0 +1,691 @@
+    /// internally. It is used for calculating the number of ticks per second for pulses.
+    ///
+    /// See [Pulse::new_with_duration()].
+    ///
+    /// [rmt_get_counter_clock]: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/rmt.html#_CPPv421rmt_get_counter_clock13rmt_channel_tP8uint32_t
+    pub fn counter_clock(&self) -> Result<Hertz, EspError> {
+        let mut ticks_hz: u32 = 0;
+        esp!(unsafe { rmt_get_counter_clock(C::channel(), &mut ticks_hz) })?;
+        Ok(ticks_hz.into())
+    }
+
+    /// Start sending the given signal without blocking.
+    ///
+    /// `signal` is captured for safety so that the user can't change the data while transmitting.
+    pub fn start<S>(&self, signal: S) -> Result<(), EspError>
+    where
+        S: Signal,
+    {
+        self.write_items(&signal, false)
+    }
+
+    /// Start sending the given signal while blocking.
+    pub fn start_blocking<S>(&self, signal: &S) -> Result<(), EspError>
+    where
+        S: Signal,
+    {
+        self.write_items(signal, true)
+    }
+
+    fn write_items<S>(&self, signal: &S, block: bool) -> Result<(), EspError>
+    where
+        S: Signal,
+    {
+        let items = signal.as_slice();
+        esp!(unsafe { rmt_write_items(C::channel(), items.as_ptr(), items.len() as i32, block,) })
+    }
+
+    /// Stop transmitting.
+    pub fn stop(&self) -> Result<(), EspError> {
+        esp!(unsafe { rmt_tx_stop(C::channel()) })
+    }
+
+    /// Stop transmitting and release the driver.
+    ///
+    /// This will return the pin and channel.
+    pub fn release(self) -> Result<(P, C), EspError> {
+        self.stop()?;
+        esp!(unsafe { rmt_driver_uninstall(C::channel()) })?;
+        Ok((self.pin, self.channel))
+    }
+
+    pub fn set_looping(&self, looping: config::Loop) -> Result<(), EspError> {
+        esp!(unsafe { rmt_set_tx_loop_count(C::channel(), looping.into()) })
+    }
+}
+
+/// Signal storage for [`Transmit`] in a format ready for the RMT driver.
+pub trait Signal {
+    fn as_slice(&self) -> &[rmt_item32_t];
+}
+
+/// Stack based signal storage for an RMT signal.
+///
+/// Use this if you know the length of the pulses ahead of time and prefer to use the stack.
+///
+/// Internally RMT uses pairs of pulses as part of its data structure. This implementation
+/// you need to [`set`][StackPairedSignal::set()] a two [`Pulse`]s for each index.
+///
+/// ```rust
+/// # use esp_idf_hal::rmt::StackPairedSignal;
+/// let p1 = Pulse::new(PinState::High, PulseTicks::new(10));
+/// let p2 = Pulse::new(PinState::Low, PulseTicks::new(11));
+/// let p3 = Pulse::new(PinState::High, PulseTicks::new(12));
+/// let p4 = Pulse::new(PinState::Low, PulseTicks::new(13));
+///
+/// let mut s = StackPairedSignal::new();
+/// s.set(0, &(p1, p2));
+/// s.set(1, &(p3, p4));
+/// ```
+#[derive(Clone)]
+pub struct StackPairedSignal<const N: usize>([rmt_item32_t; N]);
```

Comment:
Seeking for a name which will better correspond to the name of the other signal structure... How about `FixedLengthSignal`?